### ISSUE TITLE
Restructure modules

### ISSUE THREAD

**Issue by troessner** [troessner/reek#438]
I believe our current module / class organisation has some potential for improvements:

Right now the classes and modules structure i'd like to discuss in this issue look like this:

>>
lib/reek
├── core
│   ├── code_context.rb
│   ├── method_context.rb
│   ├── module_context.rb
│   ├── object_refs.rb
│   ├── singleton_method_context.rb
│   ├── smell_configuration.rb
│   ├── smell_repository.rb
│   ├── stop_context.rb
│   ├── tree_walker.rb
│   └── warning_collector.rb
├── source
│   ├── ast_node_class_map.rb
│   ├── ast_node.rb
│   ├── code_comment.rb
│   ├── reference_collector.rb
│   ├── sexp_extensions.rb
│   ├── sexp_formatter.rb
│   ├── sexp_node.rb
│   ├── source_code.rb
│   ├── source_file.rb
│   ├── source_locator.rb
│   ├── source_repository.rb
│   └── tree_dresser.rb
├── spec.rb
└── version.rb
├── examiner.rb
├── smells.rb
├── smell_warning.rb

I suggest the following changes:

- We move "smells.rb" and "smell_warning" to "./smells", "spec.rb" to "./specs", "examiner.rb" to "./core"
- We move everything in "./source" that does not start with "source_*" to "./core". Right now I feel it's highly confusing since the distinction between "./source" and "./core" seems arbitrary, at least partially.
- We move everything in "./source" that starts with "sexp_*" to "lib/reek/sexp"

I believe this would give a clearer picture of what's happening where, it would improve our name system and it would actually put responsibilities together that belong together.

What do you think?

**Comment by chastell** [comment]
I like the new grouping!

I’m not sure about the naming; I (minimally) prefer more functional ones, so maybe ‘analysis` instead of ‘core’? (I tried to come up with something better than ‘source’ and ‘sexp’, but the more I tried the more I realised they are probably good ones.)

I assume `SourceCode` will become `Source::Code`, `SexpNode` will become `Sexp::Node`, etc.?

**Comment by mvz** [comment]
Perhaps we can also simplify the require_relative calls in the specs by making sure `require 'reek'` or `require_relative '../lib/reek'` loads everything?

**Comment by mvz** [comment]
Regarding renaming `SourceCode` to `Source::Code`, etc., I would only do that for the cases where the resulting base name still makes sense on its own. I've made the mistake of creating classes like `Builder::Function` for a class that builds a function, and then ending up with very confusing code.

**Comment by chastell** [comment]
> Perhaps we can also simplify the `require_relative` calls in the specs by making sure `require 'reek'` or `require_relative '../lib/reek'` loads everything?

I had this approach in my applications, and I’m ok with it – although recently I went the other way around and try to stick to a strict ‘if a file references a non-Ruby-core class/module then it should `require` its source explicitly’ approach. I can’t say I recommend it (yet?), because it means a lot of my files start with litanies of `require` calls, but it is irritating enough for me to refactor the code so that the files stop depending on so many external classes, and I enjoy that result. Any kind of blanket ‘require everything’ approach hides (to some extent) the amount of this kind of coupling.

> Regarding renaming `SourceCode` to `Source::Code`, etc., I would only do that for the cases where the resulting base name still makes sense on its own.

Ah, a very good point! I’m not a fan of `Source::SourceCode` (as it feels redundant), but you’re fully right that readability of the places where the class/module names are actually used definitely should have precedence.

**Comment by troessner** [comment]
>> I’m not sure about the naming; I (minimally) prefer more functional ones, so maybe ‘analysis` instead of ‘core’? (I tried to come up with something better than ‘source’ and ‘sexp’, but the more I tried the more I realised they are probably good ones.)

I'm not the biggest fan of "core" as well since it's very generic. But I don't see any better name, "analysis" is kind of vague as well. I believe this suggests that we shouldn't just rename the package, but rather split it up into several ones. I can see 5 files with "context" in their name, how about leaving "core" like it is, moving "*_context*"  to "./context" and then split up the rest of "core" step by step?

>> Perhaps we can also simplify the require_relative calls in the specs by making sure require 'reek' or require_relative '../lib/reek' loads everything?

Quite frankly, I'm not the biggest fan of this. 
Here are some reasons why: http://myronmars.to/n/dev-blog/2012/12/5-reasons-to-avoid-bundler-require
So I'd rather stick with requiring files where they are used, even if that means we have duplicate require-calls to get the whole test suite running.

>> Regarding renaming SourceCode to Source::Code, etc., I would only do that for the cases where the resulting base name still makes sense on its own. I've made the mistake of creating classes like Builder::Function for a class that builds a function, and then ending up with very confusing code.

Couldn't agree more, let's stick with Source::SourceCode .:)

**Comment by chastell** [comment]
:+1: on all of the above. :)

**Comment by troessner** [comment]
Alright, I'll look into this after #417 is merged..)

**Comment by troessner** [comment]
#451 is merged - closed!
